spring-integration-gcp/src/main/java/org/springframework/integration/gcp/storage/GcsSession.java

```diff
@@ -0,0 +1,196 @@
+import java.nio.channels.Channels;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.integration.file.remote.session.Session;
+import org.springframework.util.Assert;
+import org.springframework.util.StreamUtils;
+
+/**
+ * @author João André Martins
+ */
+public class GcsSession implements Session<BlobInfo> {
+
+	private Storage gcs;
+
+	private static final Log LOGGER = LogFactory.getLog(GcsSession.class);
+
+	private static final String SEPARATOR = "/";
+
+	public GcsSession(Storage gcs) {
+		Assert.notNull(gcs, "The GCS client can't be null.");
+		this.gcs = gcs;
+	}
+
+	@Override
+	public boolean remove(String path) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(path);
+		Assert.state(tokens.length == 1 || tokens.length == 2,
+				"Path must be in the form of [bucket] or [bucket]/[blob name]");
+
+		return tokens.length == 1
+				? this.gcs.delete(tokens[0])
+				: this.gcs.delete(tokens[0], tokens[1]);
+	}
+
+	@Override
+	public BlobInfo[] list(String path) throws IOException {
+		// TODO(joaomartins): Only supports listing buckets, not folders.
+
+		Collection<BlobInfo> blobs = new ArrayList<>();
+
+		for (Blob blob : this.gcs.list(path).iterateAll()) {
+			blobs.add(blob);
+		}
+
+		return blobs.toArray(new BlobInfo[blobs.size()]);
+	}
+
+	@Override
+	public void read(String source, OutputStream outputStream) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(source);
+
+		try (OutputStream os = outputStream) {
+			os.write(this.gcs.readAllBytes(tokens[0], tokens[1]));
+		}
+	}
+
+	@Override
+	public void write(InputStream inputStream, String destination) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(destination);
+		BlobInfo gcsBlobInfo = BlobInfo.newBuilder(BlobId.of(tokens[0], tokens[1])).build();
+
+		try (InputStream is = inputStream) {
+			try (WriteChannel channel = this.gcs.writer(gcsBlobInfo)) {
+				channel.write(ByteBuffer.wrap(StreamUtils.copyToByteArray(is)));
+			}
+		}
+	}
+
+	@Override
+	public void append(InputStream inputStream, String destination) throws IOException {
+		// TODO(joaomartins): We could do compose here, but it assumes that InputStream is first copied to a
+		// GCS object that we can then "compose" to the original object.
+		throw new UnsupportedOperationException("Appending isn't supported by Google Cloud Storage.");
+	}
+
+	@Override
+	public boolean mkdir(String directory) throws IOException {
+		try {
```

Comment:
For this and rmdir, should it be checked that the directory is actually just a bucket name and not a folder with slashes? I know folders aren't like a real thing in GCS, but if some other implementations of Session allow for actual folders with slashes then people could be annoyed by GCS not supporting it and they would need to change their code. Perhaps we could make an empty nameless file in the desired folder?

Reply:
The AWS implementation assumes that a remote directory is a bucket, I think it's safe to assume the same thing on our side.
We can always change it later if we need to.

Reply:
That's correct. The story is exactly about top level bucket. Because there is no a notion like sub-dir. 
We can just mention that in the JavaDocs here to protect ourselves from similar questions in the future. In addition we may mention that in the Reference Manual if we document this `GcsSession` there at all...